### PR TITLE
ddl: support alter table add/drop/alter check constraint

### PR DESCRIPTION
**Pull request by AilinKid**
Signed-off-by: AilinKid <314806019@qq.com>

Close: https://github.com/pingcap/tidb/issues/17427, https://github.com/pingcap/tidb/issues/17426, https://github.com/pingcap/tidb/issues/17428 

### What problem does this PR solve? 
support alter table with constraint ddl action.
```
create table t(a int);

**Add constraint**:
alter table t add constraint haha check(a >1);

> Attention: check constraint will fetch the data from table to check.

**Alter constraint**:
alter table t alter constraint haha enforced;

> Attention: not enforced check constraint will allow the violated data to insert, but when change a constraint from not enforced to enforced, it will fetch the data to check constraint.

**Drop constraint**:
alter table t drop constraint haha;
```

### What is changed and how it works?
Schema State: **None** -> **Write Only** -> **Public**
**None**: Node in this state will not perceive the constraint existence, and the inserted data may violate the check constraint.

**Write Only**: Node in this state will actually see check constraint existence, but the inserted data will check the constraint rule. If broken, DML fails.

**Public**: on the way from **Write Only** to **Public**, ddl owner will fetch the old data previous to **Write Only** state, and check the data whether it's satisfied the check constraint rule. If broken, DDL fails. 

### Check List 

Tests 

 - Unit test
 - Integration test

Code changes

 - Has interface methods change

Release note

 - Write release note for bug-fix or new feature.
 - ddl: support alter table add/drop/alter check constraint

### CODE REVIEW

File: ddl/constraint.go [hunk 1]
```diff
@@ -14,15 +14,267 @@
...
+		}
+
+		tblInfo.Constraints = append(tblInfo.Constraints, constraintInfoInJob)
+		constraintInfoInMeta = constraintInfoInJob
+	}
+
+	originalState := constraintInfoInMeta.State
+	switch constraintInfoInMeta.State {
+	case model.StateNone:
+		// none -> delete only
```
Comment (djshow832) [inline ddl/constraint.go]:
```suggestion
		// none -> write only
```

File: ddl/constraint.go [hunk 2]
```diff
@@ -14,15 +14,267 @@
...
+	err = job.DecodeArgs(constraintInfo1)
+	if err != nil {
+		job.State = model.JobStateCancelled
+		return nil, nil, nil, nil, errors.Trace(err)
+	}
+	// do the double-check with constraint existence.
+	constraintInfo2 := tblInfo.FindConstraintInfoByName(constraintInfo1.Name.L)
+	if constraintInfo2 != nil {
+		if constraintInfo2.State == model.StatePublic {
+			// We already have a column with the same column name.
```
Comment (djshow832) [inline ddl/constraint.go]: Column name or constraint name?

File: ddl/constraint.go [hunk 3]
```diff
@@ -14,15 +14,267 @@
...
+			job.State = model.JobStateCancelled
+			return nil, nil, nil, nil, infoschema.ErrColumnExists.GenWithStackByArgs(constraintInfo1.Name)
+		}
+		// if not, that means constraint was in intermediate state.
+	}
+	return dbInfo, tblInfo, constraintInfo2, constraintInfo1, nil
```
Comment (djshow832) [inline ddl/constraint.go:245]: So can it be added again?

File: ddl/constraint.go [hunk 4]
```diff
@@ -77,3 +329,51 @@ func findDependedColsMapInExpr(expr ast.ExprNode) map[string]struct{} {
...
+		return errors.Trace(err)
+	}
+	defer w.sessPool.put(ctx)
+
+	// If there is any row can't pass the check expression, the add constraint action will error.
+	// It's no need to construct expression node out and pull the chunk rows through it. Here we
+	// can let the check expression restored string as the filter in where clause directly.
+	// Prepare internal SQL to fetch data from physical table under .
+	sql := fmt.Sprintf("select * from `%s`.`%s` where ", dbInfo.Name.L, tableInfo.Name.L)
+	sql = sql + " not " + constr.ExprString + " limit 1"
```
Comment (djshow832) [inline ddl/constraint.go]: Why not put it all in `Sprintf`?

Comment (djshow832) [reply ddl/constraint.go]: Why use `select *` rather than `select count(1)`?

File: ddl/constraint.go (hunk 3 above)
Comment (AilinKid) [reply ddl/constraint.go:245]: Nope
```
dbInfo, tblInfo, constraintInfoInMeta, constraintInfoInJob, err := checkAddCheckConstraint(t, job)
```
one1 is in job from the every beginning, one2 is already in meta (mid state)
one2 is prior to one1

File: ddl/constraint.go (hunk 4 above)
Comment (AilinKid) [reply ddl/constraint.go]: make sense

**Comment by AilinKid** [comment]
@djshow832 @bb7133 @imtbkcat PTAL

File: ddl/error.go [hunk 5]
```diff
@@ -206,4 +206,8 @@ var (
 	ErrColumnCheckConstraintReferOther = terror.ClassDDL.New(mysql.ErrColumnCheckConstraintReferOther, mysql.MySQLErrName[mysql.ErrColumnCheckConstraintReferOther])
 	// ErrTableCheckConstraintReferUnknown is returned when create table check constraint referring non-existing column.
 	ErrTableCheckConstraintReferUnknown = terror.ClassDDL.New(mysql.ErrTableCheckConstraintReferUnknown, mysql.MySQLErrName[mysql.ErrTableCheckConstraintReferUnknown])
+	// ErrConstraintDoesNotExist is returned for dropping a non-existent constraint.
+	ErrConstraintDoesNotExist = terror.ClassDDL.New(mysql.ErrConstraintDoesNotExist, mysql.MySQLErrName[mysql.ErrConstraintDoesNotExist])
+	// ErrCheckConstraintIsViolated is returned for violating the a existed check constraint.
```
Comment (djshow832) [inline ddl/error.go]:
```suggestion
	// ErrCheckConstraintIsViolated is returned for violating an existent check constraint.
```

**Review by djshow832 — COMMENTED** [review]
Rest LGTM

**Comment by AilinKid** [comment]
PTAL @bb7133 @imtbkcat

**Review by imtbkcat — APPROVED** [review]
LGTM

**Comment by bb7133** [comment]
LGTM, good job! @AilinKid

**Pull request merged**